### PR TITLE
[senec] Senec Home Batteries Binding initial contribution

### PR DESCRIPTION
**Pull request by vctender**
PR for new binding to integrate Senec Home devices. Those are batteries

to store solar energy to provide energy delivery through the night.

Senec Home devices provide an internal unsecured web interface to request status information.
The binding currently supports:
- Current power consumption by the home/house
- Current energy production by solar panels
- Current power limitation (reduced energy production by public energy regulations)
- Battery Fuel charge
- Battery power load

Binary/JAR
[org.openhab.binding.senechome-2.4.0-SNAPSHOT.jar](http://schwarznau.de/public/git/openhab/org.openhab.binding.senechome-2.4.0-SNAPSHOT.jar)

### CODE REVIEW

**Comment by vctender** [comment]
Known problems:
- the Unit is missing sometimes when using PaperUI (strange behaviour - reason is currently unknown)

**Comment by wborn** [comment]
Thanks for the binding PR @vctender!

> How to fix the "signed_off_by" - validation issue?

When you click the "details" button in the header above the signed_off_by requirement it will show the [requirement details](https://pullapprove.com/openhab/openhab2-addons/pull-request/3847/requirements/). 

It's failing because the author e-mail (first column) doesn't match the e-mail in the **Signed-off-by:** line.

**Comment by vctender** [comment]
I will add support for the state soon. I already have the state in my code, but it's not published to openhab yet. Maybe you can provide the mapping between the hex/dec code and the meaning?! Helps me saving time...

I know there are a lot of information available, feel free to request those, which are important.

**Comment by thopiekar** [comment]
Sure. Here it is:
image
The mapping is years old but basically works until today. Could be that the mapping slightly changed.

Is it possible that your JAR build against 2.4.0 doesn't work with 2.3.0?

**Comment by vctender** [comment]
Should work with 2.3.0, I'm running the same version. Just drop it under openhab-addons and the binding should appear under Configuration -> Bindings after some seconds.

**Comment by thopiekar** [comment]
Hmm, did that but doesn't seem to work. Maybe that is related to my setup on my Synology.

**Comment by vctender** [comment]
Hm java should be independent in that case. May try to extend logging for further details. Maybe some other users have an idea.

**Comment by thopiekar** [comment]
Hmm, might be that the add-ons directory is correctly being watched. I think it is a symlink actually.
Will try to load the add-on by restarting the service tomorrow.

**Comment by thopiekar** [comment]
Btw. shouldn't be "batteryFuelCharge" better a battery?

**Comment by thopiekar** [comment]
For example, I assigned a Text to an item called "batteryFuelCharge", which is provided by a thing representing my Senec here.
This one looks by default like this:
image
But if I set in my sitemap "icon=battery", when I get this icon for the Text, which is more logical.
image
So depending on the value the icon will change now, too.

File: addons/binding/org.openhab.binding.senechome/ESH-INF/thing/thing-types.xml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        <item-type>Number:Power</item-type>
+        <label>Current battery power</label>
+        <category>Energy</category>
+        <state readOnly="true" pattern="%.2f %unit%" />
+    </channel-type>
+    
+    <channel-type id="batteryFuelCharge">
+        <item-type>Number:Dimensionless</item-type>
+        <label>Battery fuel charge</label>
+        <category>Energy</category>
```
Comment (thopiekar) [inline addons/binding/org.openhab.binding.senechome/ESH-INF/thing/thing-types.xml]: Maybe "Battery" or something similar needs to be set here.

**Comment by vctender** [comment]
Added battery status and power limitation switch
Fixed pom's to get CI working
... link to pre-compiled JAR updated (first post)

possible release candidate

**Comment by elmario55** [comment]
Thanks for this binding,
it works fine on openhabianpi 2.3.0.1 with SENEC IES V2 5.0
I would like to see, how much of power from PV is going to grid, because i'm owner of an E-car. It would be fine to send this power to my wallbox KEBA P30.
As i have no idea about programming, i am really helpless to integrade the point "Gesamtleistung" of SENEC Information into the binding.
Help would be great!

**Comment by vctender** [comment]
I will add this information as a channel

**Comment by vctender** [comment]
Your requirement is committed and I've updated the JAR from first posting

**Comment by thopiekar** [comment]
Beautiful! Will give it a try.

Thanks!

**Comment by wborn** [comment]
Great! I've removed the tag so it will no longer be hidden due to WIP PR filters. :wink:

**Comment by elmario55** [comment]
Hi vctender,
Thanks for the great job! Now I can see the parameters of grit.
I had som problems by installing the new .jar file. I had to restart  my Openhabianpi several times to make the binding work. Actually, the binding works, but there is no entry in the paper UI addon list to install or uninstall the binding.
May be I dit something wrong

**Comment by vctender** [comment]
If you put the jar file into the addon folder, it gets automatically installed. It should appear after some seconds under Configuration -> Bindings

Please check logs for any exceptions

**Comment by elmario55** [comment]
--"If you put the jar file into the addon folder, it gets automatically 
installed. It should appear after some seconds under Configuration -> 
Bindings

Please check logs for any exceptions"

Yes, the Binding appears in Configuration->Bindings and works fine. But 
the Binding does not appear in Add-ons -> Binding, so it is not possible 
to uninstall and reinstall the Binding. If I put away the .jar-file, the 
Binding is still working.

Is this normal?

Best regards

E. Schmitt

**Comment by vctender** [comment]
@elmario55 The binding never appears in menu Addons. It is uninstalled some seconds after you've deleted the Jar-File. You can verify it by checking Configuration -> Bindings, it won't be listed anymore.
Please ensure you use the current version of JAR-File

**Comment by elmario55** [comment]
OK, thanks for this informaton, I dit'nt know this

**Comment by vctender** [comment]
@thopiekar I'm still waiting for final code review, but hope to get merged in next release

**Comment by vich-667** [comment]
Thanks for that nice plugin, I just installed it and It works great to me.  :+1: I hope this will be merged soon!
Do you also thought to add the "Day Energy Values" as Channel?

**Comment by thopiekar** [comment]
@openhab/maintainers : Are there any plans to review all the new plugins?
Would be great, if @vctender could get some feedback.

Thanks!

**Comment by matzeeg3** [comment]
Hi everybody,
i try to use this plugin, but i only recieve "err" in my sitemap.
I added the Addon by copy it, than i configure it via Paper UI.
Item generation via default.items. 
What i´m doing wrong can someone help

**Comment by vctender** [comment]
> Hi everybody,
> i try to use this plugin, but i only recieve "err" in my sitemap.
> I added the Addon by copy it, than i configure it via Paper UI.
> Item generation via default.items.
> What i´m doing wrong can someone help

You should check the logs generated by openhab for any errors. You should also share some configurations. I only use PaperUI for configuration, including item configuration.

**Comment by matzeeg3** [comment]
> > Hi everybody,
> > i try to use this plugin, but i only recieve "err" in my sitemap.
> > I added the Addon by copy it, than i configure it via Paper UI.
> > Item generation via default.items.
> > What i´m doing wrong can someone help
> 
> You should check the logs generated by openhab for any errors. You should also share some configurations. I only use PaperUI for configuration, including item configuration.

Thanks,
i was stupid ^^ :) very very good binding :) 👍

**Comment by matzeeg3** [comment]
So now i have another failure.
It all works great, but after a view houres i now have the problem, that the solar input is now multiplite with 1000 see above:

solar_openhab2
can i solve this issue?

Thanks

**Comment by vctender** [comment]
Which Senec model do you have installed? Have you checked the output when accessing your senec device directly over the internal web interface?

**Comment by matzeeg3** [comment]
> Which Senec model do you have installed? Have you checked the output when accessing your senec device directly over the internal web interface?

It is the: Senec Home V2.1 1ph
And on the Website i see everything right.
And it worked very good befor. But from now to another it has this issue.

**Comment by vctender** [comment]
Can you check Openhab Event Logs? Usually you will find there an entry for each changing value.
This will clarify if it's an error in the presentation layer or the data reader.

Website = Senec Cloud Website or the website accessed by the internal ip address?

**Comment by matzeeg3** [comment]
> Can you check Openhab Event Logs? Usually you will find there an entry for each changing value.
> This will clarify if it's an error in the presentation layer or the data reader.
> 
> Website = Senec Cloud Website or the website accessed by the internal ip address?

Thanks that you support me here. The Website is the local IP.
i attached an images with the output of the log.
show_values

**Comment by shakalandy** [comment]
@vctender do you have a working snapshot for openhab 2.5 as well? Would be interested in testing it.

File: addons/binding/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/json/SenecHomeEnergy.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ */
+public class SenecHomeEnergy implements Serializable {
+
+    /**
+     *
+     */
+    private static final long serialVersionUID = -6171687327416551070L;
+
+    private String GUI_HOUSE_POW;
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/json/SenecHomeEnergy.java]:
```suggestion
    @SerializedName("GUI_HOUSE_POW")
    public String guiHousePow;
```

We usually use camelCase for fieldnames which are not static. You could also remove the getters and setters and make the fields public. This reduces a lot of code. Remove most empty lines in files.

Also applies to the other model files.

File: addons/binding/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/json/SenecHomeEnergy.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.senechome.internal.json;
+
+import java.io.Serializable;
+
+/**
+ * @author Steven.Schwarznau
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/json/SenecHomeEnergy.java]: Please explain the class in a few words.

```suggestion
 * @author Steven Schwarznau - Initial Contribution
```

also applies to some other classes

File: addons/binding/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/json/SenecHomeEnergy.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ */
+public class SenecHomeEnergy implements Serializable {
+
+    /**
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/json/SenecHomeEnergy.java]: remove empty comments

File: addons/io/org.openhab.io.neeo/web/js/thirdparty/filesaver.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
...
+
```
Comment (J-N-K) [inline addons/io/org.openhab.io.neeo/web/js/thirdparty/filesaver.js]: I think this file was committed accidentally. Please revert

**Comment by vctender** [comment]
> > Can you check Openhab Event Logs? Usually you will find there an entry for each changing value.
> > This will clarify if it's an error in the presentation layer or the data reader.
> > Website = Senec Cloud Website or the website accessed by the internal ip address?
> 
> Thanks that you support me here. The Website is the local IP.
> i attached an images with the output of the log.
> show_values

This issue have been fixed in the source. It seems that Senec changed the response.
I've not updated the Jar file from the first post here, so please compile it yourself yet.

Due to some guidance changes you mave have to remove and re-add the thing

**Comment by vctender** [comment]
Due to some problems caused by Merge I applied the change to a clean fork state and re-created the pull https://github.com/openhab/openhab2-addons/pull/6076

**Comment by matzeeg3** [comment]
> > > Can you check Openhab Event Logs? Usually you will find there an entry for each changing value.
> > > This will clarify if it's an error in the presentation layer or the data reader.
> > > Website = Senec Cloud Website or the website accessed by the internal ip address?
> > 
> > 
> > Thanks that you support me here. The Website is the local IP.
> > i attached an images with the output of the log.
> > show_values
> 
> This issue have been fixed in the source. It seems that Senec changed the response.
> I've not updated the Jar file from the first post here, so please compile it yourself yet.
> 
> Due to some guidance changes you mave have to remove and re-add the thing

ok sorry but how

**Pull request closed without merging**